src/v/kafka/server/write_at_offset_stm.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "cluster/state_machine_registry.h"
+#include "raft/persisted_stm.h"
+#include "raft/replicate.h"
+
+namespace kafka {
+
+/**
+ * The write at offset state machine allows caller to replicate the record batch
+ * at a specific Kafka offset. The replication is only successful if the batch
+ * offset after replication matches the expected offset. Otherwise the
+ * `replicate` call will return an error. In case of an error it is the caller
+ * responsibility to reconcile with the log and retry replication of the batch
+ * with an expected offset.
+ */
+class write_at_offset_stm
+  : public raft::persisted_stm<raft::kvstore_backed_stm_snapshot> {
+public:
+    static constexpr std::string_view name = "write_at_offset";
+    /**
+     * State machine specific error codes.
+     */
+    enum class errc : int8_t {
+        success = 0,
+        invalid_offset = 1,
+        invalid_last_log_offset = 2,
+        replicate_exception = 3,
+        invalid_batch_type = 4,
+    };
+    struct errc_category final : public std::error_category {
+        const char* name() const noexcept final;
+
+        std::string message(int c) const final;
+    };
+
+    const std::error_category& error_category() noexcept;
+
+    std::error_code make_error_code(errc e) noexcept;
+
+    friend std::ostream& operator<<(std::ostream&, const errc&);
+
+    write_at_offset_stm(
+      raft::consensus* raft,
+      ss::logger& logger,
+      storage::kvstore& kvstore,
+      std::vector<model::record_batch_type> offset_translated_batches);
```

Comment:
Should this `std::vector` be a `absl::flat_hash_set`?

Also can you document what this is for and what the expected record_batch_types here are? I don't really understand the purpose here.